File: user_app/lib/features/calculators/domain/clinical_tool_evaluator.dart
import 'dart:math' as math;

import 'package:user_app/features/calculators/data/models/clinical_tool_definition.dart';

final class ClinicalToolResult {
  const ClinicalToolResult({
    required this.values,
    required this.normalizedInputs,
    this.interpretation,
    this.recommendations = const [],
    this.warnings = const [],
    this.checklist,
  });
  final Map<String, Object?> values;
  final Map<String, Object?> normalizedInputs;
  final ClinicalToolInterpretation? interpretation;
  final List<String> recommendations;
  final List<ClinicalToolMessage> warnings;
  final ClinicalToolChecklistProgress? checklist;
}

final class ClinicalToolChecklistProgress {
  const ClinicalToolChecklistProgress({
    required this.completedRequired,
    required this.totalRequired,
    required this.percentage,
    required this.complete,
    required this.needsReview,
    required this.criticalPending,
  });

  final int completedRequired;
  final int totalRequired;
  final double percentage;
  final bool complete;
  final bool needsReview;
  final List<String> criticalPending;

  Map<String, Object?> toJson() => {
    'completed_required': completedRequired,
    'total_required': totalRequired,
    'percentage': percentage,
    'complete': complete,
    'needs_review': needsReview,
    'critical_pending': criticalPending,
  };
}

final class ClinicalToolEvaluator {
  const ClinicalToolEvaluator({this.fixedNow});

  final DateTime? fixedNow;

  ClinicalToolResult evaluate(
    ClinicalToolDefinition definition,
    Map<String, Object?> input,
  ) {
    final values = <String, Object?>{...input};
    final normalizedInputs = <String, Object?>{};
    for (final field in definition.inputs) {
      if ((!values.containsKey(field.key) || values[field.key] == null) &&
          field.defaultValue != null) {
        values[field.key] = field.defaultValue;
      }
    }
    for (final field in definition.inputs) {
      final visible =
          field.visibleWhen == null ||
          _truth(_expression(field.visibleWhen!, values));
      if (visible &&
          field.required &&
          (!values.containsKey(field.key) || values[field.key] == null)) {
        throw FormatException('${field.label} is required');
      }
      final value = values[field.key];
      if (value == null) continue;
      if (value is Map && value['value'] is num) {
        final unit = value['unit']?.toString() ?? field.defaultUnit;
        values[field.key] = _convert(
          (value['value'] as num).toDouble(),
          unit,
          field.defaultUnit,
        );
        normalizedInputs[field.key] = {
          'value': values[field.key],
          'unit': field.defaultUnit,
        };
      } else {
        normalizedInputs[field.key] = value;
      }
      if (value is num && field.minimum != null && value < field.minimum!) {
        throw FormatException('${field.label} is below the minimum');
      }
      if (value is num && field.maximum != null && value > field.maximum!) {
        throw FormatException('${field.label} exceeds the maximum');
      }
    }
    for (final item in definition.calculations) {
      values[item.key] = _applyPrecision(
        _expression(item.expression, values),
        item.precision,
        item.roundingMode,
      );
    }
    final outputs = <String, Object?>{};
    for (final item in definition.outputs) {
      outputs[item.key] = _applyPrecision(
        _expression(item.value, values),
        item.precision,
        item.roundingMode,
      );
    }
    final context = <String, Object?>{...values, ...outputs};
    final interpretationByKey = {
      for (final item in definition.interpretations) item.key: item,
    };
    final warningByKey = {
      for (final item in definition.warnings) item.key: item,
    };
    ClinicalToolInterpretation? interpretation;
    final recommendations = <String>[];
    final warnings = definition.warnings
        .where(
          (warning) =>
              warning.when == null ||
              _truth(_expression(warning.when!, context)),
        )
        .toList();
    final rules = [...definition.rules]
      ..sort((a, b) {
        final byOrder = a.order.compareTo(b.order);
        return byOrder == 0 ? a.key.compareTo(b.key) : byOrder;
      });
    var stopped = false;
    for (final rule in rules) {
      if (!_truth(_expression(rule.when, context))) continue;
      for (final action in rule.actions) {
        switch (action.type) {
          case 'set_output':
            final expression = action.value;
            if (expression == null || !outputs.containsKey(action.target)) {
              throw const FormatException('Invalid set_output action');
            }
            outputs[action.target] = _expression(expression, context);
            context[action.target] = outputs[action.target];
            break;
          case 'add_interpretation':
            final item = interpretationByKey[action.target];
            if (item == null) {
              throw FormatException('Unknown interpretation: ${action.target}');
            }
            interpretation ??= item;
            _appendUnique(recommendations, item.recommendations);
            break;
          case 'add_recommendation':
            final item = interpretationByKey[action.target];
            if (item == null) {
              throw FormatException(
                'Unknown recommendation source: ${action.target}',
              );
            }
            _appendUnique(recommendations, item.recommendations);
            break;
          case 'add_warning':
          case 'escalate':
            final item = warningByKey[action.messageKey];
            if (item == null) {
              throw FormatException('Unknown warning: ${action.messageKey}');
            }
            if (!warnings.any((warning) => warning.key == item.key)) {
              warnings.add(item);
            }
            break;
          case 'stop':
            stopped = true;
            break;
        }
      }
      if (stopped || rule.stop) break;
    }
    final interpretations = [...definition.interpretations]
      ..sort((a, b) {
        final byOrder = a.order.compareTo(b.order);
        return byOrder == 0 ? a.key.compareTo(b.key) : byOrder;
      });
    for (final item in interpretations) {
      if (_truth(_expression(item.when, context))) {
        interpretation ??= item;
        _appendUnique(recommendations, item.recommendations);
      }
    }
    ClinicalToolChecklistProgress? checklist;
    if (definition.toolType == 'checklist') {
      final required = definition.inputs
          .where((field) => field.type == 'checklist_item' && field.required)
          .toList(growable: false);
      bool responseComplete(Object? value) =>
          value == true ||
          value is num ||
          (value is String && value.isNotEmpty) ||
          (value is List && value.isNotEmpty);
      final completed = required
          .where((field) => responseComplete(input[field.key]))
          .length;
      final criticalPending = required
          .where(
            (field) => field.critical && !responseComplete(input[field.key]),
          )
          .map((field) => field.key)
          .toList(growable: false);
      final needsReview = definition.completion.requireReview;
      checklist = ClinicalToolChecklistProgress(
        completedRequired: completed,
        totalRequired: required.length,
        percentage: required.isEmpty
            ? 100
            : (completed / required.length * 10000).round() / 100,
        complete:
            definition.completion.mode == 'all_required' &&
            completed == required.length &&
            !needsReview &&
            criticalPending.isEmpty,
        needsReview: needsReview,
        criticalPending: List.unmodifiable(criticalPending),
      );
    }
    return ClinicalToolResult(
      values: outputs,
      normalizedInputs: Map.unmodifiable(normalizedInputs),
      interpretation: interpretation,
      recommendations: List.unmodifiable(recommendations),
      warnings: List.unmodifiable(warnings),
      checklist: checklist,
    );
  }

  Object? _expression(
    ClinicalToolExpression expression,
    Map<String, Object?> values,
  ) {
    List<Object?> args() => expression.args
        .map((item) => _expression(item, values))
        .toList(growable: false);
    switch (expression.op) {
      case 'literal':
        return expression.value;
      case 'field':
        return values[expression.field];
      case 'add':
        return args().fold<double>(0, (sum, value) => sum + _number(value));
      case 'subtract':
        final value = args();
        return _number(value[0]) - _number(value[1]);
      case 'multiply':
        return args().fold<double>(
          1,
          (product, value) => product * _number(value),
        );
      case 'divide':
        final value = args();
        final divisor = _number(value[1]);
        if (divisor == 0) throw const FormatException('Division by zero');
        return _number(value[0]) / divisor;
      case 'power':
        final value = args();
        return math.pow(_number(value[0]), _number(value[1])).toDouble();
      case 'min':
        return args().map(_number).reduce(math.min);
      case 'max':
        return args().map(_number).reduce(math.max);
      case 'abs':
        return _number(_expression(expression.args.first, values)).abs();
      case 'greater_than':
        final value = args();
        return _number(value[0]) > _number(value[1]);
      case 'greater_than_or_equal':
        final value = args();
        return _number(value[0]) >= _number(value[1]);
      case 'less_than':
        final value = args();
        return _number(value[0]) < _number(value[1]);
      case 'less_than_or_equal':
        final value = args();
        return _number(value[0]) <= _number(value[1]);
      case 'equal':
        final value = args();
        return value[0] == value[1];
      case 'not_equal':
        final value = args();
        return value[0] != value[1];
      case 'and':
        return expression.args.every(
          (item) => _truth(_expression(item, values)),
        );
      case 'or':
        return expression.args.any((item) => _truth(_expression(item, values)));
      case 'not':
        return !_truth(_expression(expression.args.first, values));
      case 'if':
        return _truth(_expression(expression.args[0], values))
            ? _expression(expression.args[1], values)
            : _expression(expression.args[2], values);
      case 'in':
        final value = args();
        return value.skip(1).contains(value.first);
      case 'round':
        final value = _number(_expression(expression.args.first, values));
        return _applyPrecision(
          value,
          expression.precision ?? 0,
          expression.roundingMode,
        );
      case 'now':
        return (fixedNow ?? DateTime.now()).toUtc().toIso8601String();
      case 'date_difference':
        final value = args();
        final from = _parseDate(value[0]);
        final to = _parseDate(value[1]);
        final difference = to.difference(from);
        return switch (expression.dateUnit) {
          'minutes' => difference.inMinutes,
          'hours' => difference.inHours,
          'weeks' => difference.inMilliseconds / 86400000 / 7,
          'months' => _calendarMonths(from, to),
          'years' => _calendarYears(from, to),
          _ => difference.inMilliseconds / 86400000,
        };
      case 'date_add':
        final value = args();
        final source = _parseDate(value[0]);
        final amount = _number(value[1]);
        if (amount.truncateToDouble() != amount) {
          throw const FormatException('date_add requires an integer amount');
        }
        final result = switch (expression.dateUnit) {
          'days' => source.add(Duration(days: amount.toInt())),
          'weeks' => source.add(Duration(days: amount.toInt() * 7)),
          'months' => DateTime.utc(
            source.year,
            source.month + amount.toInt(),
            source.day,
          ),
          'years' => DateTime.utc(
            source.year + amount.toInt(),
            source.month,
            source.day,
          ),
          _ => throw const FormatException('Unsupported date_add unit'),
        };
        return '${result.year.toString().padLeft(4, '0')}-${result.month.toString().padLeft(2, '0')}-${result.day.toString().padLeft(2, '0')}';
      case 'convert_unit':
        return _convert(
          _number(_expression(expression.args.first, values)),
          expression.fromUnit ?? '',
          expression.toUnit ?? '',
        );
      default:
        throw UnsupportedError(
          'Unsupported clinical tool operation: ${expression.op}',
        );
    }
  }

  double _number(Object? value) {
    if (value is num) return value.toDouble();
    throw const FormatException('Expected a numeric value');
  }

  DateTime _parseDate(Object? value) {
    final text = value.toString();
    return DateTime.parse(
      text.length == 10 ? '${text}T00:00:00Z' : text,
    ).toUtc();
  }

  int _calendarMonths(DateTime from, DateTime to) {
    var months = (to.year - from.year) * 12 + to.month - from.month;
    if (to.day < from.day) months--;
    return months;
  }

  int _calendarYears(DateTime from, DateTime to) {
    var years = to.year - from.year;
    if (to.month < from.month ||
        (to.month == from.month && to.day < from.day)) {
      years--;
    }
    return years;
  }

  Object? _applyPrecision(Object? value, int? precision, String? mode) {
    if (precision == null || value is! num || !value.isFinite) return value;
    final scale = math.pow(10, precision).toDouble();
    final scaled = value.toDouble() * scale;
    final rounded = switch (mode) {
      'floor' => scaled.floorToDouble(),
      'ceil' => scaled.ceilToDouble(),
      'truncate' => scaled.truncateToDouble(),
      'half_even' => _roundHalfEven(scaled),
      _ => scaled.sign * (scaled.abs() + 0.5).floorToDouble(),
    };
    return rounded / scale;
  }

  double _roundHalfEven(double value) {
    final lower = value.floorToDouble();
    final fraction = value - lower;
    if (fraction == 0.5) return lower.toInt().isEven ? lower : lower + 1;
    return value.roundToDouble();
  }

  bool _truth(Object? value) => value == true;

  void _appendUnique(List<String> target, Iterable<String> values) {
    for (final value in values) {
      if (!target.contains(value)) target.add(value);
    }
  }

  double _convert(double value, String from, String to) {
    if (from.isEmpty || to.isEmpty || from == to) return value;
    const factors = <String, double>{
      'kg': 1,
      'g': 0.001,
      'mg': 0.000001,
      'mcg': 0.000000001,
      'lb': 0.45359237,
      'm': 1,
      'cm': 0.01,
      'mm': 0.001,
      'in': 0.0254,
      'ft': 0.3048,
      'L': 1,
      'mL': 0.001,
      'weeks': 10080,
      'days': 1440,
      'hours': 60,
      'minutes': 1,
    };
    if (from == 'celsius' && to == 'fahrenheit') return value * 9 / 5 + 32;
    if (from == 'fahrenheit' && to == 'celsius') return (value - 32) * 5 / 9;
    final source = factors[from];
    final target = factors[to];
    if (source == null || target == null) {
      throw FormatException('Unsupported unit conversion: $from to $to');
    }
    return value * source / target;
  }
}

File: dashboard/lib/clinical-tool-evaluator.ts
import type { ClinicalToolDefinition, ClinicalToolExpression } from "@/services/clinical-tool.service"

export type ClinicalToolChecklistProgress = { completed_required: number; total_required: number; percentage: number; complete: boolean; needs_review: boolean; critical_pending: string[] }
export type ClinicalToolPreviewResult = { values: Record<string, unknown>; normalizedInputs: Record<string, unknown>; interpretation?: string; recommendations: string[]; warnings: string[]; checklist?: ClinicalToolChecklistProgress }

const unitFactors: Record<string, number> = { kg: 1, g: .001, mg: .000001, mcg: .000000001, lb: .45359237, m: 1, cm: .01, mm: .001, ft: .3048, in: .0254, L: 1, mL: .001, weeks: 10080, days: 1440, hours: 60, minutes: 1 }
function convertUnit(value: number, from = "", to = ""): number {
  if (!from || !to || from === to) return value
  if (from === "celsius" && to === "fahrenheit") return value * 9 / 5 + 32
  if (from === "fahrenheit" && to === "celsius") return (value - 32) * 5 / 9
  if (!(from in unitFactors) || !(to in unitFactors)) throw new Error(`Unsupported unit conversion: ${from} to ${to}`)
  return value * unitFactors[from] / unitFactors[to]
}

function applyPrecision(value: unknown, precision?: number, mode = "half_up"): unknown {
  if (precision === undefined || typeof value !== "number" || !Number.isFinite(value)) return value
  const scale = 10 ** precision
  const scaled = value * scale
  let rounded: number
  switch (mode) {
    case "floor": rounded = Math.floor(scaled); break
    case "ceil": rounded = Math.ceil(scaled); break
    case "truncate": rounded = Math.trunc(scaled); break
    case "half_even": {
      const lower = Math.floor(scaled)
      const fraction = scaled - lower
      rounded = fraction === 0.5 ? (lower % 2 === 0 ? lower : lower + 1) : Math.round(scaled)
      break
    }
    default: rounded = Math.sign(scaled) * Math.floor(Math.abs(scaled) + 0.5)
  }
  return rounded / scale
}

function evaluate(expression: ClinicalToolExpression, values: Record<string, unknown>, now: string): unknown {
  const args = () => (expression.args ?? []).map((item) => evaluate(item, values, now))
  switch (expression.op) {
    case "literal": return expression.value
    case "field": return values[expression.field ?? ""]
    case "add": return args().reduce<number>((sum, value) => sum + Number(value), 0)
    case "subtract": { const [a, b] = args(); return Number(a) - Number(b) }
    case "multiply": return args().reduce<number>((total, value) => total * Number(value), 1)
    case "divide": { const [a, b] = args(); if (Number(b) === 0) throw new Error("Division by zero"); return Number(a) / Number(b) }
    case "power": { const [a, b] = args(); return Number(a) ** Number(b) }
    case "min": return Math.min(...args().map(Number))
    case "max": return Math.max(...args().map(Number))
    case "abs": return Math.abs(Number(args()[0]))
    case "greater_than": { const [a, b] = args(); return Number(a) > Number(b) }
    case "greater_than_or_equal": { const [a, b] = args(); return Number(a) >= Number(b) }
    case "less_than": { const [a, b] = args(); return Number(a) < Number(b) }
    case "less_than_or_equal": { const [a, b] = args(); return Number(a) <= Number(b) }
    case "equal": { const [a, b] = args(); return a === b }
    case "not_equal": { const [a, b] = args(); return a !== b }
    case "and": return (expression.args ?? []).every((item) => Boolean(evaluate(item, values, now)))
    case "or": return (expression.args ?? []).some((item) => Boolean(evaluate(item, values, now)))
    case "not": return !Boolean(evaluate((expression.args ?? [])[0], values, now))
    case "if": return Boolean(evaluate((expression.args ?? [])[0], values, now)) ? evaluate((expression.args ?? [])[1], values, now) : evaluate((expression.args ?? [])[2], values, now)
    case "in": { const [needle, ...haystack] = args(); return haystack.some((item) => Object.is(item, needle)) }
    case "round": { const [value] = args(); return applyPrecision(Number(value), expression.precision ?? 0, expression.rounding_mode) }
    case "now": return now
    case "date_difference": {
      const [fromValue, toValue] = args(); const milliseconds = Date.parse(String(toValue)) - Date.parse(String(fromValue)); const days = milliseconds / 86_400_000
      if (!Number.isFinite(days)) throw new Error("Invalid date_difference operands")
      if (expression.date_unit === "months" || expression.date_unit === "years") {
        const from = new Date(String(fromValue)); const to = new Date(String(toValue))
        if (expression.date_unit === "months") {
          let months = (to.getUTCFullYear() - from.getUTCFullYear()) * 12 + to.getUTCMonth() - from.getUTCMonth()
          if (to.getUTCDate() < from.getUTCDate()) months--
          return months
        }
        let years = to.getUTCFullYear() - from.getUTCFullYear()
        if (to.getUTCMonth() < from.getUTCMonth() || (to.getUTCMonth() === from.getUTCMonth() && to.getUTCDate() < from.getUTCDate())) years--
        return years
      }
      switch (expression.date_unit) { case "minutes": return days * 1440; case "hours": return days * 24; case "weeks": return days / 7; default: return days }
    }
    case "date_add": {
      const [dateValue, amountValue] = args()
      const source = String(dateValue)
      const amount = Number(amountValue)
      if (!Number.isInteger(amount)) throw new Error("date_add requires an integer amount")
      const date = new Date(source.length === 10 ? `${source}T00:00:00Z` : source)
      if (!Number.isFinite(date.getTime())) throw new Error("Invalid date_add date")
      switch (expression.date_unit) {
        case "days": date.setUTCDate(date.getUTCDate() + amount); break
        case "weeks": date.setUTCDate(date.getUTCDate() + amount * 7); break
        case "months": date.setUTCMonth(date.getUTCMonth() + amount); break
        case "years": date.setUTCFullYear(date.getUTCFullYear() + amount); break
        default: throw new Error("Unsupported date_add unit")
      }
      return date.toISOString().slice(0, 10)
    }
    case "convert_unit": return convertUnit(Number(args()[0]), expression.from_unit, expression.to_unit)
    default: throw new Error(`Unsupported preview operation: ${expression.op}`)
  }
}

export function previewClinicalTool(definition: ClinicalToolDefinition, input: Record<string, unknown>, options: { fixedNow?: string } = {}): ClinicalToolPreviewResult {
  const now = options.fixedNow ?? new Date().toISOString()
  const values = { ...input }
  const normalizedInputs: Record<string, unknown> = {}
  const knownInputs = new Set(definition.inputs.map((field) => field.key))
  for (const key of Object.keys(input)) if (!knownInputs.has(key)) throw new Error(`Unknown input: ${key}`)
  for (const field of definition.inputs) {
    if ((values[field.key] === undefined || values[field.key] === null) && field.default !== undefined && field.default !== null) values[field.key] = field.default
  }
  for (const field of definition.inputs) {
    const candidate = values[field.key]
    const visible = !field.visible_when || Boolean(evaluate(field.visible_when, values, now))
    if (visible && field.required && (candidate === undefined || candidate === null || candidate === "")) throw new Error(`Required input: ${field.key}`)
    if (candidate === undefined || candidate === null) continue
    if (candidate && typeof candidate === "object" && "value" in candidate) {
      const measurement = candidate as { value: unknown; unit?: string }
      values[field.key] = convertUnit(Number(measurement.value), measurement.unit, field.default_unit)
      normalizedInputs[field.key] = { value: values[field.key], unit: field.default_unit }
    } else {
      normalizedInputs[field.key] = candidate
    }
  }
  for (const calculation of definition.calculation) values[calculation.key] = applyPrecision(evaluate(calculation.expression, values, now), calculation.precision, calculation.rounding_mode)
  const output: Record<string, unknown> = {}
  for (const item of definition.outputs) output[item.key] = applyPrecision(evaluate(item.value, values, now), item.precision, item.rounding_mode)
  const context = { ...values, ...output }
  const interpretationByKey = new Map(definition.interpretations.map((item) => [item.key, item]))
  const warningByKey = new Map((definition.warnings ?? []).map((item) => [item.key, item]))
  let interpretation: (typeof definition.interpretations)[number] | undefined
  const recommendations: string[] = []
  const warnings = (definition.warnings ?? []).filter((item) => !item.when || Boolean(evaluate(item.when, context, now))).map((item) => item.text)
  const append = (items: string[]) => { for (const item of items) if (!recommendations.includes(item)) recommendations.push(item) }
  for (const rule of [...definition.rules].sort((a, b) => a.order - b.order || a.key.localeCompare(b.key))) {
    if (!Boolean(evaluate(rule.when, context, now))) continue
    let stop = false
    for (const action of rule.actions) {
      if (action.type === "set_output" && action.target && action.value) { output[action.target] = evaluate(action.value, context, now); context[action.target] = output[action.target] }
      if ((action.type === "add_interpretation" || action.type === "add_recommendation") && action.target) { const item = interpretationByKey.get(action.target); if (!item) throw new Error(`Unknown interpretation: ${action.target}`); if (action.type === "add_interpretation") interpretation ??= item; append(item.recommendations) }
      if ((action.type === "add_warning" || action.type === "escalate") && action.message_key) { const item = warningByKey.get(action.message_key); if (!item) throw new Error(`Unknown warning: ${action.message_key}`); if (!warnings.includes(item.text)) warnings.push(item.text) }
      if (action.type === "stop") stop = true
    }
    if (stop || rule.stop) break
  }
  for (const item of [...definition.interpretations].sort((a, b) => a.order - b.order || a.key.localeCompare(b.key))) {
    if (Boolean(evaluate(item.when, context, now))) { interpretation ??= item; append(item.recommendations) }
  }
  let checklist: ClinicalToolChecklistProgress | undefined
  if (definition.tool_type === "checklist") {
    const required = definition.inputs.filter((field) => field.type === "checklist_item" && field.required)
    const completeResponse = (value: unknown) => value === true || typeof value === "number" || (typeof value === "string" && value.length > 0) || (Array.isArray(value) && value.length > 0)
    const completed = required.filter((field) => completeResponse(input[field.key])).length
    const criticalPending = required.filter((field) => field.critical && !completeResponse(input[field.key])).map((field) => field.key)
    const needsReview = Boolean(definition.completion.require_review)
    checklist = {
      completed_required: completed,
      total_required: required.length,
      percentage: required.length === 0 ? 100 : Math.round(completed / required.length * 10_000) / 100,
      complete: definition.completion.mode === "all_required" && completed === required.length && !needsReview && criticalPending.length === 0,
      needs_review: needsReview,
      critical_pending: criticalPending,
    }
  }
  return { values: output, normalizedInputs, interpretation: interpretation?.label, recommendations, warnings, checklist }
}
